### DISCUSSION TITLE
Best way to implement RBAC?

### DISCUSSION THREAD

**Discussion by sturatcliffe** [supabase/supabase#346] [Questions] [answered]
Hey,

Is there a recommended way of implementing RBAC? Before I start implementing UI's and such I'm just trying to decide on an approach, so this is what I've tried:

- Modified the ```auth.users``` table to change a users role from 'authenticated' to 'admin'. This seems to correctly change the role returned with ```supabase.auth.user()``` which is what I wanted, and I believe it changes the ```auth.role()``` within Postgres for RLS. However, it also stops all queries working by default - I assume there's some built-in RLS to only allow the 'authenticated' or 'anon' roles?

- I already have a public users table which is populated by a trigger off the ```auth.users``` table, so I added a role column which I can query independently of ```supabase.auth.user()```, and this should also work with RLS policies. However, the downside is I have a RLS policy to allow users to modify their own personal info in this public users table, meaning they'd also be free to modify their role. I haven't managed to find any way of limiting an RLS policy to allow modification of specific columns within a table, so I'm assuming as it's called **Row** Level Security, it applies to the whole row or nothing?

- I can only think of one other option of having a secondary user table just for mapping user id's to roles, which is locked down with RLS so it can't be modified. The only downside to this that I can see is how clean (or not as the case may be) the code becomes, as we end up needing to query multiple tables to get all the info we need to correctly identify and check access for a user, or add a view to allow us to simplify the query.

Are there any other obvious (and nicer) ways of achieving RBAC? Or a simple way of fixing the authenticated role being the only one which can make queries by default?

Cheers!

**Comment by steve-chavez** [comment]
Hey @sturatcliffe 

Some pointers here.

> so I'm assuming as it's called Row Level Security, it applies to the whole row or nothing?

Yes, that's right. For disallowing modification of certain columns you can grant privileges at the column level, for example:

```sql
-- only list columns that will be modified("name" and "other" in this case)
GRANT UPDATE(name, other) ON my_users TO authenticated;
```

> I already have a public users table which is populated by a trigger off the auth.users table

For this case you might consider a VIEW instead of a table. Simple VIEWs are [auto-updatable](https://www.postgresql.org/docs/current/sql-createview.html#SQL-CREATEVIEW-UPDATABLE-VIEWS), meaning you can INSERT/UPDATE/DELETE on them. With a view you can save the need for a trigger.

What I would recommend, is to make your users VIEW only SELECTable for your roles - not GRANT any INSERT/UPDATE privilege on it.

To allow users to modify their own data, you can use a stored function with the [security definer option](https://postgrest.org/en/v7.0.0/schema_structure.html#security-definer). An example:

```sql
create or replace function modify_my_data(new_name text) returns users as $$
  update users set name = new_name where id = auth.uid(); -- a user can only modify it's own data
$$ language sql SECURITY DEFINER

-- make sure to REVOKE public access(all roles) from this function(all functions are like this by default)
revoke execute on function modify_my_data(text) from public;

-- and only grant execute for your authenticated role
grant execute on function modify_my_data(text) to authenticated;
```

Then you can call the function through our [rpc interface](https://supabase.io/docs/client/rpc). That way, your table is protected from modification - it can only be done through the function.

Let me know if that helps!

**Reply by steve-chavez** [reply]
@wadclapp Modifying the `auth` schema is not recommended since is a Supabase internal schema and we'll change it during upgrades. I'd suggest you add another table and then reference the `auth.users` table through a foreign key.

**Comment by raimille1** [comment]
great question @sturatcliffe ! With respect to the first point you make, did you end up playing with your own roles? Or stayed with the anon/authenticated ones? Curious as I'm looking into a similar approach. You can then modify RLS to actually give `my-role` users your own access policies.

**Comment by humleflue** [comment]
Here is my take on it: https://lasse-skaalum.medium.com/implementing-rbac-with-next-js-and-supabase-5dd20adaaeba

**Reply by RobSchilderr** [reply]
Any idea if this is on the roadmap?

**Comment by thorwebdev** [comment, marked as answer]
We've released some official guidance around custom claims & RBAC using the new Auth Hooks approach: https://supabase.com/docs/guides/auth/custom-claims-and-role-based-access-control-rbac

**Reply by iAbhinav** [reply]
> We've released some official guidance around custom claims & RBAC using the new Auth Hooks approach: https://supabase.com/docs/guides/auth/custom-claims-and-role-based-access-control-rbac

Hey @thorwebdev 

I have implemented the suggested RBAC and it is working.
The problem is, when the admin removes a row from the user_roles table, user's JWT still contains the custom claim until user logs out and logs in again.

Is there a way to update a user's role without logging him out?

**Reply by sturatcliffe** [reply]
Hey @iAbhinav,

This is just how JWT's work unfortunately, once they've been issued, you can't invalidate them or change them without issuing a new one.

In a traditional server based backend, you would have to store some blacklist of JWT's somehow which have had claims updated, and then utilise some kind of middleware to ensure blacklisted users can't perform any queries until they've obtained a new JWT. This obviously adds a level of overhead as this middleware would be running on every request to your backend.

However, I have no idea how you would go about this if you're querying supabase directly from your frontend, without a backend proxy in the middle.

Hope this helps.
